runtime/src/iree/hal/utils/deferred_work_queue.c

```diff
@@ -1039,38 +1032,51 @@ static iree_status_t iree_hal_hip_pending_queue_actions_issue_execution(
                ? IREE_HAL_COMMAND_BUFFER_MODE_UNVALIDATED
                : 0);
       IREE_RETURN_AND_END_ZONE_IF_ERROR(
-          z0, iree_hal_hip_device_create_stream_command_buffer(
-                  action->device, mode, IREE_HAL_COMMAND_CATEGORY_ANY,
-                  /*binding_capacity=*/0, &stream_command_buffer));
+          z0, device_interface->_vtable->create_command_buffer_for_deferred(
+                  device_interface, mode, IREE_HAL_COMMAND_CATEGORY_ANY,
+                  &stream_command_buffer))
       IREE_RETURN_AND_END_ZONE_IF_ERROR(
           z0, iree_hal_resource_set_insert(action->resource_set, 1,
                                            &stream_command_buffer));
+
       IREE_RETURN_AND_END_ZONE_IF_ERROR(
           z0, iree_hal_deferred_command_buffer_apply(
                   command_buffer, stream_command_buffer, binding_table));
-      iree_hal_hip_stream_notify_submitted_commands(stream_command_buffer);
+      command_buffer = stream_command_buffer;
+      release = true;
+    }
+
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, device_interface->_vtable->submit_command_buffer(device_interface,
+                                                             command_buffer));
+
+    if (release) {
       // The stream_command_buffer is going to be retained by
       // the action->resource_set and deleted after the action
       // completes.
-      iree_hal_resource_release(stream_command_buffer);
+      iree_hal_resource_release(command_buffer);
     }
   }
+
   IREE_TRACE_ZONE_END(z_dispatch_command_buffers);
 
-  hipEvent_t completion_event = NULL;
-  // Last record hipEvent_t signals in the dispatch stream.
+  void* completion_event = NULL;
```

Comment:
use typedefs for this and other types (it can be typedef to void*, but void* is not readable)